### PR TITLE
Do not use global variable for containerslist

### LINKED ISSUE #468: API Operator is mixing containers of several APIs when several APIs reconciled at the same time
**Description:**

When the API Operator has to reconcile several APIs at the same time (for example when the API Operator pod restarts), it will add several containers to micro-gateways pods.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

k8s-api-operator 1.2.1

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

- Create 3 APIs in the same namespace, with some time between creations
- Wait for mgw pods to run 
- you get:
```
clientlists-api-cb4566dd7-prbpm    1/1     Running   0          5s
clients-api-79ff65b884-qppvj          1/1     Running   0          40s
users-api-6cc6477fd6-q85kh         1/1     Running   0          120s
```
- delete the operator pod and wait 
- you get:
```
clientlists-api-cb4566dd7-prbpm   1/3     CrashLoopBackOff    3          40s
clients-api-79ff65b884-qppvj         1/1     Running                     0          40s
users-api-6cc6477fd6-q85kh         1/2     CrashLoopBackOff   3          40s
```

I'm not sure to understand why you are managing a global list of containers in https://github.com/wso2/k8s-api-operator/blob/8ee6b0a0a65027f6a55955ac9bfbeffcca141a6b/api-operator/pkg/mgw/deployment.go#L53, but I think that you should pass the newly created container and not the complete list at: https://github.com/wso2/k8s-api-operator/blob/8ee6b0a0a65027f6a55955ac9bfbeffcca141a6b/api-operator/pkg/mgw/deployment.go#L196

**Related Issues:**

### PR DESCRIPTION
**Pull request by feloy**
## Purpose

Resolves #468 

## Goals

Create a local instance of the containersList for each call to the reconcile loop, instead of using the same global one.

Use local instance of ContainersList

## Documentation

N/A

### CODE REVIEW

File: api-operator/pkg/mgw/deployment.go [hunk 1]
```diff
@@ -49,22 +50,18 @@ const (
...
 	initContainerList := make([]corev1.Container, 0, 2)
-	ContainerList = &initContainerList
+	return &initContainerList
 }
 
-func AddContainers(containers *[]corev1.Container) {
-	*ContainerList = append(*ContainerList, *containers...)
+func AddContainers(containerList *[]corev1.Container, containers *[]corev1.Container) {
+	*containerList = append(*containerList, *containers...)
 }
```
Comment (renuka-fernando) [inline api-operator/pkg/mgw/deployment.go:60]: Shall we remove this function as well, it is just appending two slices

File: api-operator/pkg/endpoints/sidecar.go [hunk 2]
```diff
@@ -38,7 +39,7 @@ const (
 
 var logger = log.Log.WithName("endpoints.sidecar")
 
-func AddSidecarContainers(client *client.Client, apiNamespace string, sidecarEpNames *map[string]bool) error {
+func AddSidecarContainers(client *client.Client, apiNamespace string, sidecarEpNames *map[string]bool, containersList *[]corev1.Container) error {
 	containerList := make([]corev1.Container, 0, len(*sidecarEpNames))
```
Comment (renuka-fernando) [inline api-operator/pkg/endpoints/sidecar.go:43]: Shall we remove this `containerList` variable  and append sidecar containers directly.

**Comment by pmartin-lvmh** [comment]
Thanks @renuka-fernando for the review. I've made some changes to limit the scope and mutability of the containers list

**Comment by feloy** [comment]
Thanks @renuka-fernando , I've made some changes to reduce the scope and mutability of the containers list variable

**Review by renuka-fernando — APPROVED** [review]

**Review by pubudu538 — APPROVED** [review]

**Comment by pubudu538** [comment]
Thank you for your contribution! We highly appreciate that.

**Pull request merged**